### ISSUE TITLE
Progressive word extension on triple/more click

### ISSUE THREAD

**Issue by martinetd** [kovidgoyal/kitty#967]
Hi and thanks for kitty!
I'm just testing waters here, if you're ok with the idea I might look at implementing this myself at some point™

One of the feature I have only ever seen in urxvt is the multiple click selection that is 'smart' -- they do it the opposite way you do with cut chars instead of include char but it can work as well.

Currently, double-click stops as soon as it sees a char that is not alphanum or whitelisted, and then a third click will grab the whole line.

What I'd like to see (as an option perhaps) is to extend the selection in the word past one non-allowed char on both side, but still stopping at spaces. And when there is no extension left possible, grab the whole line.

For example, `don't`, double-clicking on the `don` will grab don, then one more click would grab `don't`, then a fourth click would grab the line.
Another example would be `"foo"`, double-clicking will select `foo`, triple-clicking will select `"foo"` with the quotes, four clicks will get the whole line.
(a third use-case would be links that include cut-characters, but we can use the hint feature for that)

urxvt extends by one everytime so something like `example'with'quotes` on the first word would stop at `example'with` on third click, then the whole thing on four and the line on fifth but I am not particularly picky about this behaviour, it would still be useful if it were the whole word on third and line one fourth for example.

Similarily, more clicks after the full line is selected toggle between the last two in urxvt, but the kitty behaviour of just cancelling selection works for me.

(FWIW, while testing this, I noticed a bug when double-clicking in a while space (e.g. large spaces between files in ls output) then extending the selection with right-click upwards, the end of the selection slowly crawls to the right... I think it might make more sense to pick the whole line from the start when double-clicking in whitespace)

Code-wise it should be pretty easy, instead of counting in multi_click just always call the same screen_select_extend_range and have that function gobble up a few chars; if both edges already are blanks then grab whole line.

This is, honestly, pretty situational but I get frustrated fairly regularly without it ; what do you think?

**Comment by kovidgoyal** [comment]
I dont think I want to change the default behavior, I am not a big fan of the concept of clicks having different effects based on how many clicks happened in the near past. It makes clicking a somewhat chancy affair. However, I am willing to have an option to turn on this behavior. PRs are welcome. I am closing this bug report, but if you wish to discuss more, feel free to comment.

**Comment by martinetd** [comment]
Thanks for the reply - I'll probably look at this at some point after the secondary scrollback buffer if you're open to featurebloat :)

**Comment by kovidgoyal** [comment]
I'm fine with feature bloat as long as the features are implemented well and dont affect main-line performance. You just need to read the default kitty.conf to see evidence of that :)

**Comment by martinetd** [comment]
I've looked "a bit" at this over the past few days and stumbled on many small hitches with selection, would like advices on what you consider a fix and what would be better handled separately. Feel free to wait until you have more time to answer, there's no hurry.

 * right click selection extension updates on mouse release, I believe it should update on mouse-down and movement so you can push right click, hold and move around like a normal left click in cell mode
 * right click right now updates based on the initial selection; but I believe it should act based on the current state of the selection and only change one edge at a time e.g. with current selection represented by pipes, if you start with `this is |a sample| text` and click somewhere on text you get `this is |a sample te|xt` then clicking on this gives `th|is is a sample| text` and I believe it should be `th|is is a sample te|xt`. I could make that optional but I'd rather not have to.
 * double-clicking a blank then moving mouse up will make the blank crawl right till next word, definitely a bug, screen_update_selection is too greedy in EXPAND_WORD case.
 * wide chars are not handled at all, that's easily considered a bug.. Not difficult to fix however.

Up to this point I have working implementations that I'm fiddling with, and am mostly waiting to not have multiple PRs open in parallel and need to strip the "smart expansion" option back off at first, so please don't bother fixing any of this yourself unless you really want to
(sorry; playing with new options was more attractive than cleanup of the scrollback PR! I'll need to do some though, even if you do not feel strongly about style there are definitely places I'll want to tweak a bit myself before a merge now I've got a better picture of the codebase).

 * In the code I see a few `if (BLANK_CHAR == 0)` checks but this is a DEFINE, are these obsolete checks? Should new code check for this kind of things?
 * I haven't looked at this part yet but I'd like to handle unicode spaces as well eventually (spaces are different from non-alphanumeric unicode), but I'm a bit dreading having to make another `is_word_char()`-like function for blanks... Probably later in a separate PR.
 * Now on to the main thing I'm worried, that is blank/new line handling:

basically right now we have:
 - a single EXPAND_LINE line does not contain any new line, spaces at the start/end of line are stripped;
 - similarly, multiple lines selected with EXPAND_LINE do not contain a leading nor trailing new line; the first line has leading spaces stripped, the last line has trailing spaces stripped.
 - multiple lines selected without EXPAND_LINE by single clicking on the right of the text and just moving up/down include a leading newline, no trailing newline.
 - visually, the end of the line is never highlighted; spaces at the end of a line are highlighted when they are selected.

What I'd like to see is none of these include a leading newline (this rarely make sense); concretely with single cell selection I'd only like to include the first newline if the character just after the last character of the previous line is "selected". It makes the newline impossible to select for a line that goes until the end of the row, unless a character from the next row is also selected, but that is the same as right now.
I also would want whole line to include the trailing newline (and thus trailing spaces) but that could be... another option as this changes the behaviour? I'm stil not sure about leading spaces, I find stripping them to be kind of useful, but a bit treacherous; this option might disable this.

Finally, to make this more obvious, I would want the trailing part of the line to be highlighted as a sign of "the new line" is selected.

Whatever the outcome, this would definitely be in its own PR, later; this is a bigger change than the other fixes.

**Comment by martinetd** [comment]
* Also something I didn't really notice until recently, selecting something and then scrolling up through the selection feels really slow, and that seems artificial (CPU usage isn't high at all, although it's a bit weird because the speed isn't constant).
Some multiplier option here to do a few lines at a time might be good perhaps?

**Comment by kovidgoyal** [comment]
On Tue, Sep 18, 2018 at 06:38:55PM -0700, Dominique Martinet wrote:
> I've looked "a bit" at this over the past few days and stumbled on many small hitches with selection, would like advices on what you consider a fix and what would be better handled separately. Feel free to wait until you have more time to answer, there's no hurry.
> 
>  * right click selection extension updates on mouse release, I believe it should update on mouse-down and movement so you can push right click, hold and move around like a normal left click in cell mode

Sounds reasonable to me. Feel free to send a separate PR for this.

>  * right click right now updates based on the initial selection; but I believe it should act based on the current state of the selection and only change one edge at a time e.g. with current selection represented by pipes, if you start with `this is |a sample| text` and click somewhere on text you get `this is |a sample te|xt` then clicking on this gives `th|is is a sample| text` and I believe it should be `th|is is a sample te|xt`. I could make that optional but I'd rather not have to.

This seems wrong to me. The idea of the current behavior is that if you
started the selection in "select words mode" or "select lines mode"
extending it should continue in the same mode. That seems like the
"natural" behavior to me. 


>  * double-clicking a blank then moving mouse up will make the blank crawl right till next word, definitely a bug, screen_update_selection is too greedy in EXPAND_WORD case.

Probably, I 'd need to looka t this in more detail.

>  * wide chars are not handled at all, that's easily considered a bug.. Not difficult to fix however.

I though I did handle them, but I could be wrong, I am not a big mouse
user so this part of kitty is a little neglected.

> 
> Up to this point I have working implementations that I'm fiddling with, and am mostly waiting to not have multiple PRs open in parallel and need to strip the "smart expansion" option back off at first, so please don't bother fixing any of this yourself unless you really want to

No worries -- I am totally swamped at the moment.

> (sorry; playing with new options was more attractive than cleanup of the scrollback PR! I'll need to do some though, even if you do not feel strongly about style there are definitely places I'll want to tweak a bit myself before a merge now I've got a better picture of the codebase).

Sure, I totally understand wanting the code you write to meet certain
standards.

> 
>  * In the code I see a few `if (BLANK_CHAR == 0)` checks but this is a DEFINE, are these obsolete checks? Should new code check for this kind of things?

This is mainly a way to document where the assumption that blank chars
are 0 (the other candidate is 32) is. Historically kitty used 32, but
that has various drawbacks. When I changed it, I used this technique so
I could easily grep the code to see the impact.

>  * I haven't looked at this part yet but I'd like to handle unicode spaces as well eventually (spaces are different from non-alphanumeric unicode), but I'm a bit dreading having to make another `is_word_char()`-like function for blanks... Probably later in a separate PR.

Fine by me, mouse selection is not exactly performance critical, so
knock yourself out. unicode data is generated directly from the unicode
standard via the gen-wcwidth.py script.

>  * Now on to the main thing I'm worried, that is blank/new line handling:
> 
> basically right now we have:
>  - a single EXPAND_LINE line does not contain any new line, spaces at the start/end of line are stripped;
>  - similarly, multiple lines selected with EXPAND_LINE do not contain a leading nor trailing new line; the first line has leading spaces stripped, the last line has trailing spaces stripped.
>  - multiple lines selected without EXPAND_LINE by single clicking on the right of the text and just moving up/down include a leading newline, no trailing newline.
>  - visually, the end of the line is never highlighted; spaces at the end of a line are highlighted when they are selected.
> 
> What I'd like to see is none of these include a leading newline (this rarely make sense); concretely with single cell selection I'd only like to include the first newline if the character just after the last character of the previous line is "selected". It makes the newline impossible to select for a line that goes until the end of the row, unless a character from the next row is also selected, but that is the same as right now.
> I also would want whole line to include the trailing newline (and thus trailing spaces) but that could be... another option as this changes the behaviour? I'm stil not sure about leading spaces, I find stripping them to be kind of useful, but a bit treacherous; this option might disable this.
> 
> Finally, to make this more obvious, I would want the trailing part of the line to be highlighted as a sign of "the new line" is selected.
> 
> Whatever the outcome, this would definitely be in its own PR, later; this is a bigger change than the other fixes.

This requires more careful thought than I have the bandwidth for at the
moment. 

Also please see #949 it's likely your PRs will clash with that. If you
dont want to wait for @jscheid to finish his work, you can probably take
over the PR, fix the issues and get it merged, assuming @jscheid has no
objection.

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Comment by kovidgoyal** [comment]
On Tue, Sep 18, 2018 at 07:19:13PM -0700, Dominique Martinet wrote:
> * Also something I didn't really notice until recently, selecting something and then scrolling up through the selection feels really slow, and that seems artificial (CPU usage isn't high at all, although it's a bit weird because the speed isn't constant).
> Some multiplier option here to do a few lines at a time might be good perhaps?

With the caveat that I am not a heavy mouse user, it seems to be the
usual way this is done in GUI apps is to control the speed by moving the
mouse further outside the window while dragging. Since kitty is often
used fullscreen, this may not be suitable (no place to move the mouse
further).

**Comment by martinetd** [comment]
Thanks for the answers!

>> * right click right now updates based on the initial selection; but I believe it should act based on the current state of the selection and only change one edge at a time e.g. with current selection represented by pipes, if you start with `this is |a sample| text` and click somewhere on text you get `this is |a sample te|xt` then clicking on this gives `th|is is a sample| text` and I believe it should be `th|is is a sample te|xt`. I could make that optional but I'd rather not have to.
>
> This seems wrong to me. The idea of the current behavior is that if you
started the selection in "select words mode" or "select lines mode"
extending it should continue in the same mode. That seems like the
"natural" behavior to me.

Sure, that's something else entirly. With word selection for example you'd have:
`this is a |sample| text example` to `this is a |sample text| example` to `this |is a sample| text example` by clicking on text then on is sequentially, I'd like the final result to be `this |is a sample text| example` instead.

I'm not touching the logic of word/line expansion here

> End of lines discussion

No worry, this can wait.

> #949 

Argh, this does conflict quite a bit ;)
That'll be a good way to rework what I've done in a progressive, easier to ingest way though so I guess it's not so bad; I'll wait until he's done as I've been slacking at work way too much lately and have stuff to finish this week -- and I want to finish the scrollback PR first anyway so there is time. Although if you need help fixing the bottom-right corner bug @jscheid please ask.

> scrollback speed through selection

Yeah, almost-fullscreen is the way to go, so I think it has to always be the same speed. A multiplier in the config file sounds like the simplest way to deal with this to me.
An alternative is mouse movement, e.g. selecting and putting the mouse out of the window goes at a given speed, but moving the mouse around (including left and right) speeds it up - that is distance is not how far from the window you are but how much the mouse moved since you got out.. I've seen that done so I assume we get input events for that but that might be less obvious; I'd rather go with simple multiplier.

**Comment by kovidgoyal** [comment]
On Tue, Sep 18, 2018 at 08:30:01PM -0700, Dominique Martinet wrote:
> > scrollback speed through selection
> 
> Yeah, almost-fullscreen is the way to go, so I think it has to always be the same speed. A multiplier in the config file sounds like the simplest way to deal with this to me.

The rpoblem with that is, you often want to scroll quickly through large
parts of text and then slow down to recognize a stopping point more
precisely, which a multiplier does not allow.

> An alternative is mouse movement, e.g. selecting and putting the mouse out of the window goes at a given speed, but moving the mouse around (including left and right) speeds it up - that is distance is not how far from the window you are but how much the mouse moved since you got out.. I've seen that done so I assume we get input events for that but that might be less obvious; I'd rather go with simple multiplier.

Yeah you get input events if a drag started in your window. But how
would one slow down in this scheme? Distance from initial exit point?

**Comment by martinetd** [comment]
> scrollback speed through selection

Hm, to be honest with as few as 2k lines in the main scrollback I don't think I'll want to slow down much, or rather I won't go very far in the first place -- a simple multiplier is just something I'd be comfortable with to not wait too long yet already is slow enough. But that is my personal use case, someone with a huge real scrollback might want something more complete.

For the 'movement-as-distance' what I've seen is that movement triggers a short time-limited burst of speed, stopping movement goes back to slow (or the screen does not scroll at all if the pointer is not moving, and you only get as many lines worth as what you moved; this would be counter-intuitive to many users I think).
I guess distance from exit point could also work, we usually have some leeway in width

**Comment by kovidgoyal** [comment]
On Tue, Sep 18, 2018 at 08:54:51PM -0700, Dominique Martinet wrote:
> For the 'movement-as-distance' what I've seen is that movement triggers a short time-limited burst of speed, stopping movement goes back to slow (or the screen does not scroll at all if the pointer is not moving, and you only get as many lines worth as what you moved; this would be counter-intuitive to many users I think).
> I guess distance from exit point could also work, we usually have some leeway in width

Distance from exit point seems the best way to me. But for something
like this it would be good to get feedback from more people. If/when you
want to implement it, I suggest opening an issue requesting feedback
first.

**Comment by blueyed** [comment]
@martinetd 
Do you have anything that I could test already?
I am interested in the urxvt-selection like behavior (smart expansion of selection on 3+ clicks).

**Comment by blueyed** [comment]
For reference: doc and code of urvt's selection extension: https://github.com/exg/rxvt-unicode/blob/0767fe71f667f5be4e8967ca6ea27e2b14c19daf/src/perl/selection#L5-L44.
